traits/observers/_filtered_trait_observer.py

```diff
@@ -0,0 +1,186 @@
+    filter : callable(str, CTrait) -> boolean
+        A callable that receives the name of a trait and the corresponding
+        trait definition. The returned boolean indicates whether the trait is
+        observed. In order to remove an existing observer with the equivalent
+        filter, the filter callables must compare equally.
+    """
+
+    def __init__(self, notify, filter):
+        self.notify = notify
+        self.filter = filter
+
+    def __hash__(self):
+        """ Return a hash of this object."""
+        return hash((type(self).__name__, self.notify, self.filter))
+
+    def __eq__(self, other):
+        """ Return true if this observer is equal to the given one."""
+        return (
+            type(self) is type(other)
+            and self.notify == other.notify
+            and self.filter == other.filter
+        )
+
+    def iter_observables(self, object):
+        """ Yield the instance traits matching the filter given. If trait
+        definition cannot be found on the object, raise an error.
+
+        Parameters
+        ----------
+        object: object
+            Object provided by the ``iter_objects`` methods from another
+            observers or directly by the user.
+
+        Yields
+        ------
+        IObservable
+
+        Raises
+        ------
+        ValueError
+            If the given object cannot be handled by this observer.
+        """
+        try:
+            name_to_ctrait = object.traits()
+        except AttributeError:
+            raise ValueError(
+                "Unable to obtain trait definitions from {!r}".format(object)
+            )
+
+        for name, ctrait in name_to_ctrait.items():
+            if self.filter(name, ctrait):
+                yield object._trait(name, 2)
+
+    def iter_objects(self, object):
+        """ Yield the values of the traits that match the given filter. The
+        values will be given to the next observer(s) following this one in an
+        ObserverGraph.
+
+        If a value has not been set as an instance attribute, i.e. absent in
+        ``object.__dict__``, this observer will skip it. This is to avoid
+        evaluating default initializers while adding observers. When the
+        default is defined, a change event will trigger the maintainer to
+        add/remove notifiers for the next observers.
+
+        Note that ``Undefined``, ``Uninitialized`` and ``None`` values are also
+        skipped, as they are unavoidable filled values that contain no further
+        attributes to be observed by any other observers.
+
+        Parameters
+        ----------
+        object: object
+            Object provided by the ``iter_objects`` methods from another
+            observers or directly by the user.
+
+        Yields
+        ------
+        value : object
+        """
+        try:
+            name_to_ctrait = object.traits()
+        except AttributeError:
+            raise ValueError(
```

Comment:
How likely are we to encounter this in practice? Do we expect to catch this `ValueError` programmatically at some other point in the code?

If this is usually going to be the result of a coding error, then I think we should just let the `AttributeError` propagate and not re-raise with a different type.

On the other hand, if this is something that we're going to need to catch to deal with it elsewhere in the observer machinery, we may want to consider a more specific exception type.

Reply:
Good point. The interface allows this method raise whatever exception deems appropriate, nothing catches the exception. I will remove this.

Reply:
> nothing catches the exception

Sorry. This is not strictly true, exceptions are captured but is then re-raised:
https://github.com/enthought/traits/blob/fc45d75c1068058929cd3e11a3c0b68063a3164a/traits/observers/_observe.py#L91-L99

Reply:
I missed answering this question:

> How likely are we to encounter this in practice?

It depends. It is intended for capturing programming error by the user.

> If this is usually going to be the result of a coding error

i.e. this condition is true.

For example, if this observer follows an `ListItemObserver` and there is an item in the list that isn't an instance of `HasTraits`, this observer should complain.

Reply:
Okay; let's remove this. There are cases where one wants to catch an exception only to immediately re-raise as another type of exception (e.g., when conforming to some API that makes use of a particular exception type for control-flow), but I don't think this is one of them.